### PR TITLE
import/export accounts

### PR DESCRIPTION
**Pull request by zelig**
#541 
https://github.com/ethereum/go-ethereum/wiki/Managing-your-accounts
- cli: add passwordfile flag
- cli: change unlock flag only takes account
- cli: with unlock you are prompted for password or use passfile with password flag
- cli: unlockAccount used in normal client start (run) and accountExport
- cli: getPassword used in accountCreate and accountImport
- accounts: Manager.Import, Manager.Export
- crypto: SaveECDSA (to complement LoadECDSA) to save to file
- crypto: NewKeyFromECDSA added (used in accountImport and New = generated constructor)

### CODE REVIEW

File: cmd/ethereum/main.go [hunk 1]
```diff
@@ -83,11 +83,62 @@ The output of this command is supposed to be machine-readable.
...
+
+Exports the given account's private key into keyfile using the canonical EC format.
+The account needs to be unlocked, if it is not the user is prompted for a passphrase to unlock it.
+For non-interactive use, the password can be specified with the --unlock flag:
+
+    ethereum --unlock <passwrdfile> account export <address> <keyfile>
+
+Note:
+Since you can directly copy your encrypted accounts to another ethereum instance, this import/export mechanism is not needed when you transfer an account between nodes.
+					`,
```
Comment (fjl) [inline cmd/ethereum/main.go]: Please wrap these strings at column 80.

File: cmd/utils/flags.go [hunk 2]
```diff
@@ -104,7 +104,13 @@ var (
 	}
 	UnlockedAccountFlag = cli.StringFlag{
 		Name:  "unlock",
-		Usage: "Unlock a given account untill this programs exits (address:password)",
+		Usage: "unlock the account given until this program exits (prompts for password).",
+		Value: "",
+	}
+	PasswordFileFlag = cli.StringFlag{
+		Name:  "password",
+		Usage: "Password used when saving a new account and unlocking an existing account. If you create a new account make sure you remember this password.",
```
Comment (fjl) [inline cmd/utils/flags.go]: Please make this description shorter.

File: crypto/crypto.go [hunk 3]
```diff
@@ -139,6 +139,11 @@ func LoadECDSA(file string) (*ecdsa.PrivateKey, error) {
 	return ToECDSA(buf), nil
 }
 
+// SaveECDSA saves a secp256k1 private key from the given file.
+func SaveECDSA(file string, key *ecdsa.PrivateKey) error {
+	return common.WriteFile(file, FromECDSA(key))
```
Comment (fjl) [inline crypto/crypto.go]: I think this function provides no additional value. The caller might as well just use `WriteFile` directly.

File: cmd/ethereum/main.go [hunk 4]
```diff
@@ -83,11 +83,62 @@ The output of this command is supposed to be machine-readable.
 					Action: accountList,
 					Name:   "list",
 					Usage:  "print account addresses",
+					Description: `
+
+`,
```
Comment (fjl) [inline cmd/ethereum/main.go]: Please remove this description.

File: crypto/crypto.go (hunk 3 above)
Comment (obscuren) [inline crypto/crypto.go]: I agree

**Comment by obscuren** [comment]
Change original behaviour of `--unlock`

File: accounts/account_manager.go [hunk 5]
```diff
@@ -101,6 +101,17 @@ func (am *Manager) firstAddr() ([]byte, error) {
 	return addrs[0], nil
 }
 
+func (am *Manager) getKey(addr []byte, keyAuth string) (*crypto.Key, error) {
+	if len(addr) == 0 {
```
Comment (fjl) [inline accounts/account_manager.go]: This is not a good idea. `Unlock(nil, ...)` should never work.
This change is also unrelated to the account import/export.

File: crypto/crypto.go (hunk 3 above)
Comment (fjl) [inline crypto/crypto.go]: Looking further, `common.WriteFile` is another one of those shortcuts that could be replaced with `ioutil.WriteFile`.
For writing private keys, it's probably in the interest of the user to create the file with non world-readable permissions.

    ioutil.WriteFile(file, crypto.FromECDSA(key), 0600)

File: cmd/ethereum/main.go [hunk 6]
```diff
@@ -218,22 +289,30 @@ func execJSFiles(ctx *cli.Context) {
...
-		if len(split) != 2 {
-			utils.Fatalf("Illegal 'unlock' format (address:password)")
-		}
-		am := eth.AccountManager()
-		// Attempt to unlock the account
-		err := am.Unlock(common.FromHex(split[0]), split[1])
-		if err != nil {
-			utils.Fatalf("Unlock account failed '%v'", err)
+		if account == "coinbase" {
+			account = ""
```
Comment (fjl) [inline cmd/ethereum/main.go]: Use `am.Coinbase` here to get the coinbase address, then unlock with that.

File: accounts/account_manager.go (hunk 5 above)
Comment (obscuren) [inline accounts/account_manager.go]: An error needs to be returned if an invalid `addr` is passed.

File: accounts/account_manager.go [hunk 7]
```diff
@@ -208,3 +219,35 @@ func zeroKey(k *ecdsa.PrivateKey) {
 		b[i] = 0
 	}
 }
+
+func (am *Manager) Export(path string, addr []byte, keyAuth string) error {
+	key, err := am.keyStore.GetKey(addr, keyAuth)
+	if err != nil {
+		return err
+	}
+	return crypto.SaveECDSA(path, key.PrivateKey)
```
Comment (obscuren) [inline accounts/account_manager.go]: Get rid of this in favour of the suggested method below

File: crypto/crypto.go (hunk 3 above)
Comment (zelig) [inline crypto/crypto.go]: agreed. will keep saveECDSA for this exactly. 
eliminate common.writeFile everywhere?

File: accounts/account_manager.go (hunk 7 above)
Comment (zelig) [inline accounts/account_manager.go]: whenever theres a load, theres a save. know this and never be sorry.
but writing with proper permissions is great idea.

File: cmd/ethereum/main.go (hunk 6 above)
Comment (zelig) [inline cmd/ethereum/main.go]: absolutely better, thx.

**Comment by fjl** [comment]
No, please don't make this change now. We can phase it out slowly.

File: crypto/crypto.go (hunk 3 above)
Comment (fjl) [inline crypto/crypto.go]: No need to do it now. Just don't introduce another use.

**Comment by fjl** [comment]
We can merge this once you remove `crypto.SaveECDSA`.

**Comment by fjl** [comment]
Also maybe squash the changes together if you want.

File: accounts/account_manager.go (hunk 7 above)
Comment (obscuren) [inline accounts/account_manager.go]: I've given it some thought and I believe this is actually right. It enforces a permission mode and leaves less room for errors or mistakes. Leave it in.

File: cmd/ethereum/main.go [hunk 8]
```diff
@@ -255,30 +342,97 @@ func accountList(ctx *cli.Context) {
...
+		utils.Fatalf("account address must be given as first argument")
+	}
+	keyfile := ctx.Args().Get(1)
+	if len(keyfile) == 0 {
+		utils.Fatalf("keyfile must be given as second argument")
+	}
+	am := utils.GetAccountManager(ctx)
+	auth := unlockAccount(ctx, am, account)
+
+	err := am.Export(keyfile, common.FromHex(account), auth)
```
Comment (obscuren) [inline cmd/ethereum/main.go]: We shouldn't export private keys directly to disk. Export encrypted keys only. Maybe supply an additional tool to actually export the private key.

People tend to forget they exported something to file. Ending up with your HDD compromised will lead to your private keys being compromised. We should try to do everything we can do minimise the risk. Supplying an additional tool means "look, you can do it, we don't recommend it, but feel free to use this tool."

Importing is a different story. Would it be possible to check the format of the key and assume either "encrypted key" or "unencrypted key"?

Comment (Gustav-Simonsson) [inline cmd/ethereum/main.go]: The current files have different JSON schema for encrypted and unencrypted, but it's not very clean to use compared to a proper header. @fjl and I discussed this before and talked about adding a common format with a header that specifies whether what follows is encrypted or not.

Comment (zelig) [inline cmd/ethereum/main.go]: so currently there are 2 consistent options which both make sense:

1 leave everything there as is. 
- we support unencrypted keys anyway as part of the cli with a simple flag then no point in worrying about unencrypted export for now. 
- Its all developers anyway. and it is hidden in a subcommand. You really need to be tech savvy to do this.

2 full paranoid mode:
- remove export 
- remove plain text key-store 
- remove --unencrypted-keys flag

testers can setup a coinbase and unlock it all non-interactively with the following commands:

```
ethereum -datadir /tmp/eth/42 -logfile /dev/null -password <(echo -n notsosecret) account new 
ethereum -datadir /tmp/eth/42 -logfile /dev/null -port 30342  js <(echo 'console.log(admin.nodeInfo().NodeUrl)') > enode
ethereum -datadir /tmp/eth/42 -logfile 42.log -port 30342- password <(echo -n notsosecret) -unlock coinbase .....<further flags to run the node>
```
and therefore test team won't even need a custom built client

Comment (obscuren) [inline cmd/ethereum/main.go]: Option number `2` :-)
